src/router.rs

```diff
@@ -27,14 +35,124 @@ impl ComponentRequest {
+        let response: StatusResponse = serde_json::from_str(&body)?;
+        debug!("{:?}", response.active_components);
+        let mut component_list = Vec::new();
+
+        for component in &response.active_components {
+            let component_name = format!("{}/{}", component.id.path.user, component.id.path.repo);
+            debug!("{}", component_name);
+            component_list.push(component_name);
+        }
+
+        debug!("{:?}", component_list);
+
+        Ok(component_list)
+    }
+}
+
+#[derive(Debug)]
+pub struct RequestForwarder {
+    workers: Vec<Arc<WorkerNode>>,
+    components_map: RwLock<HashMap<String, Arc<WorkerNode>>>,
+}
+
+impl RequestForwarder {
+    pub fn new() -> Self {
+        let mut workers = Vec::new();
+        let components_map = RwLock::new(HashMap::new());
+
+        // If loading from the environment variable fails, there was a user error and we should
+        // bail
+        
+        let worker_string = match env::var("V9_WORKERS") {
+            Ok(value) => value,
+            Err(e) => panic!("No V9_WORKERS env variable set: {:?}", e),
+        };
+
+        let env_workers: Vec<&str> = worker_string.split(';').collect();
+
+        for worker in &env_workers {
+            workers.push(Arc::new(WorkerNode::new(worker.to_string())));
+        }
+
+        let request_forwarder = Self {
+            workers,
+            components_map,
+        };
+
+        // Scan through server list to get initial active components
+        if let Err(e) = request_forwarder.update_workers() {
+            error!("Initial worker update failed: {:?}", e);
+        }
+
+        request_forwarder
+    }
+
+    pub fn update_workers(&self) -> Result<(), RouterError> {
+        self.components_map.write().clear();
+        for worker in &self.workers {
+            let component_list = worker.get_component_list()?;
+            for component_name in &component_list {
+                debug!("{}", component_name);
+                self.components_map
+                    .write()
+                    .insert(component_name.to_string(), Arc::clone(worker));
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn forward_request(&self, request: ComponentRequest) -> Result<Response<Body>, RouterError> {
+        let component_name = format!("{}/{}", request.user, request.repo);
+        if !self.components_map.read().contains_key(&component_name) {
+            self.update_workers()?;
+        }
+
+        debug!("Forwarding request to: {:?}", component_name);
+
+        if !self.components_map.read().contains_key(&component_name) {
+            let body = Body::from("Requested component not found\n");
+            warn!("Component to forward to not found: {:?}", component_name);
+            return Ok(Response::builder().status(404).body(body).unwrap());
+        }
+
+        let mut url = format!(
+            "{}/sl/{}/{}",
+            self.components_map.read().get(&component_name).unwrap().url,
```

Comment:
You need to be more careful with the RwLock here. You take it to check if the entry is in the map, release it, retake it, and then assume the entry is in the map. This is a race condition